tests/test_model_building.py

```diff
@@ -0,0 +1,164 @@
+        .with_columns(previous=pl.col("daily").shift(1).over("geography"))
+        .with_columns(date=pl.col("date").str.strptime(pl.Date, "%Y-%m-%d"))
+    )
+
+    frame = iup.IncidentUptakeData(frame)
+
+    return frame
+
+
+def test_extract_starting_conditions(frame):
+    """
+    Use the last date for each grouping factor
+    """
+    output = iup.LinearIncidentUptakeModel.extract_starting_conditions(
+        frame, ("geography",)
+    )
+
+    assert output.sort("geography").equals(
+        pl.DataFrame(
+            {
+                "geography": ["USA", "PA"],
+                "last_date": [dt.date(2020, 1, 21), dt.date(2020, 1, 21)],
+                "last_daily": [4.0 / 7, 0.4 / 7],
+                "last_elapsed": [21, 21],
+                "last_cumulative": [8.0, 0.8],
+            }
+        ).sort("geography")
+    )
+
+
+def test_extract_standards(frame):
+    """
+    Make a dictionary with as many keys as provided variables,
+    and two values per sub-dictionary
+    """
+    output = iup.LinearIncidentUptakeModel.extract_standards(
+        frame, ("estimate", "elapsed")
+    )
+
+    assert len(output) == 2
+    assert len(output["estimate"]) == 2
+
+
+def test_fit(frame):
+    """
+    Model should fit a line to two points, giving a 'perfect' fit
+    """
+
+    output = iup.LinearIncidentUptakeModel()
+    output = output.fit(frame, ("geography",))
+
+    assert output.model.score(output.x, output.y) == 1.0
+
+
+def test_build_scaffold_handles_no_groups():
+    """
+    Set up a data frame with 5 rows and 7 columns if no grouping factors are given
+    """
+    start = pl.DataFrame(
+        {
+            "last_date": dt.date(2020, 1, 31),
+            "last_daily": 0.1,
+            "last_elapsed": 31,
+            "last_cumulative": 4.0,
+            "last_interval": 1,
+        }
+    )
+    start_date = dt.date(2020, 2, 1)
+    end_date = dt.date(2020, 2, 29)
+    interval = "7d"
+    group_cols = None
+
+    output = iup.LinearIncidentUptakeModel.build_scaffold(
+        start, start_date, end_date, interval, group_cols
+    )
+
+    assert output.shape[0] == 5
+    assert output.shape[1] == 7
+
+
+def test_build_scaffold_handles_groups():
+    """
+    Set up a data frame with 10 rows and 8 columns if a grouping factor with 2 values is given
+    """
+    start = pl.DataFrame(
+        {
+            "last_date": dt.date(2020, 1, 31),
+            "last_daily": 0.1,
+            "last_elapsed": 31,
+            "last_cumulative": 4.0,
+            "last_interval": 1,
+        }
+    ).join(pl.DataFrame({"geography": ["USA", "PA"]}), how="cross")
+    start_date = dt.date(2020, 2, 1)
+    end_date = dt.date(2020, 2, 29)
+    interval = "7d"
+    group_cols = ("geography",)
+
+    output = iup.LinearIncidentUptakeModel.build_scaffold(
+        start, start_date, end_date, interval, group_cols
+    )
+
+    assert output.shape[0] == 10
+    assert output.shape[1] == 8
+
+
+def test_project_sequentially():
```

Comment:
love this test. are there any other trivial or edge cases that are easy to test?

Reply:
Likely so - I'll give it some thought but merge what I have so far.